File: python/src/fluke_3540/tariff.py
"""TOU tariff calculator — translate Wh_total into $$ cost.

Supports a simple peak/off-peak schedule. Peak hours are a list of
``(start_hour, end_hour)`` pairs in 24-hour wall time UTC (we follow the
record timestamps, which are UTC). Any record outside the peak windows
is billed at the off-peak rate.

Future: per-weekday schedules, holidays, tiered demand charges. Out of
scope for v0.4.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Iterable

from .parser import FIELDS, Record


_FIELD_INDEX = {f.name: f.index for f in FIELDS}


@dataclass(frozen=True)
class Tariff:
    currency: str = "USD"
    peak_rate: float = 0.0        # $/kWh
    offpeak_rate: float = 0.0     # $/kWh
    peak_hours: tuple[tuple[int, int], ...] = ()  # ((start_h, end_h),)
    # end_h is exclusive; a value < start_h means the window wraps midnight.

    def is_peak(self, hour: int) -> bool:
        for start, end in self.peak_hours:
            if start == end:
                continue
            if start < end:
                if start <= hour < end:
                    return True
            else:  # wraps midnight
                if hour >= start or hour < end:
                    return True
        return False


def compute_cost(records: Iterable[Record], tariff: Tariff) -> dict:
    """Walk records, bucket Wh_total by peak/off-peak, return cost breakdown.

    Returns a dict with:
      currency, peak_kwh, offpeak_kwh, peak_cost, offpeak_cost,
      imported_cost, exported_cost, net_cost,
      peak_imported_kwh, peak_exported_kwh,
      offpeak_imported_kwh, offpeak_exported_kwh
    """
    if "Wh_total" not in _FIELD_INDEX:
        return _empty(tariff)
    wh_idx = _FIELD_INDEX["Wh_total"]
    pk_imp = pk_exp = op_imp = op_exp = 0.0
    for r in records:
        wh = r.floats[wh_idx]
        if not wh:
            continue
        hour = r.start.hour
        is_peak = tariff.is_peak(hour)
        if wh > 0:
            if is_peak:
                pk_imp += wh
            else:
                op_imp += wh
        else:
            if is_peak:
                pk_exp += wh
            else:
                op_exp += wh
    pk_imp_k = pk_imp / 1000.0
    pk_exp_k = pk_exp / 1000.0
    op_imp_k = op_imp / 1000.0
    op_exp_k = op_exp / 1000.0
    peak_imp_cost = pk_imp_k * tariff.peak_rate
    peak_exp_cost = pk_exp_k * tariff.peak_rate
    off_imp_cost = op_imp_k * tariff.offpeak_rate
    off_exp_cost = op_exp_k * tariff.offpeak_rate
    imported_cost = peak_imp_cost + off_imp_cost
    exported_cost = peak_exp_cost + off_exp_cost
    return {
        "currency": tariff.currency,
        "peak_kwh":             pk_imp_k + pk_exp_k,
        "offpeak_kwh":          op_imp_k + op_exp_k,
        "peak_imported_kwh":    pk_imp_k,
        "peak_exported_kwh":    pk_exp_k,
        "offpeak_imported_kwh": op_imp_k,
        "offpeak_exported_kwh": op_exp_k,
        "peak_cost":            peak_imp_cost + peak_exp_cost,
        "offpeak_cost":         off_imp_cost + off_exp_cost,
        "imported_cost":        imported_cost,
        "exported_cost":        exported_cost,
        "net_cost":             imported_cost + exported_cost,
    }


def _empty(tariff: Tariff) -> dict:
    keys = (
        "peak_kwh offpeak_kwh peak_imported_kwh peak_exported_kwh "
        "offpeak_imported_kwh offpeak_exported_kwh peak_cost offpeak_cost "
        "imported_cost exported_cost net_cost"
    ).split()
    return {"currency": tariff.currency, **{k: 0.0 for k in keys}}

File: web/tariff.js
// TOU tariff calculator — JS port of python/.../tariff.py.

function fieldIndex(spec, name) {
  const f = spec.fields.find((x) => x.name === name);
  return f ? f.index : -1;
}

/**
 * @typedef {object} Tariff
 * @property {string} currency
 * @property {number} peakRate     - $/kWh
 * @property {number} offpeakRate  - $/kWh
 * @property {Array<[number, number]>} peakHours - [[startH, endH], ...] in UTC
 */

export function isPeak(tariff, hour) {
  for (const [start, end] of tariff.peakHours ?? []) {
    if (start === end) continue;
    if (start < end) {
      if (start <= hour && hour < end) return true;
    } else {
      if (hour >= start || hour < end) return true;
    }
  }
  return false;
}

export function emptyCost(currency = 'USD') {
  return {
    currency,
    peakKwh: 0, offpeakKwh: 0,
    peakImportedKwh: 0, peakExportedKwh: 0,
    offpeakImportedKwh: 0, offpeakExportedKwh: 0,
    peakCost: 0, offpeakCost: 0,
    importedCost: 0, exportedCost: 0, netCost: 0,
  };
}

export function computeCost(records, spec, tariff) {
  if (!tariff) return emptyCost();
  const whIdx = fieldIndex(spec, 'Wh_total');
  if (whIdx < 0) return emptyCost(tariff.currency);
  let pkImp = 0, pkExp = 0, opImp = 0, opExp = 0;
  for (const r of records) {
    const wh = r.floats[whIdx];
    if (!wh) continue;
    const hour = new Date(r.startMs).getUTCHours();
    const peak = isPeak(tariff, hour);
    if (wh > 0) {
      if (peak) pkImp += wh; else opImp += wh;
    } else {
      if (peak) pkExp += wh; else opExp += wh;
    }
  }
  const pkImpK = pkImp / 1000, pkExpK = pkExp / 1000;
  const opImpK = opImp / 1000, opExpK = opExp / 1000;
  const peakImpCost = pkImpK * tariff.peakRate;
  const peakExpCost = pkExpK * tariff.peakRate;
  const offImpCost = opImpK * tariff.offpeakRate;
  const offExpCost = opExpK * tariff.offpeakRate;
  return {
    currency: tariff.currency,
    peakKwh: pkImpK + pkExpK,
    offpeakKwh: opImpK + opExpK,
    peakImportedKwh: pkImpK,
    peakExportedKwh: pkExpK,
    offpeakImportedKwh: opImpK,
    offpeakExportedKwh: opExpK,
    peakCost: peakImpCost + peakExpCost,
    offpeakCost: offImpCost + offExpCost,
    importedCost: peakImpCost + offImpCost,
    exportedCost: peakExpCost + offExpCost,
    netCost: (peakImpCost + offImpCost) + (peakExpCost + offExpCost),
  };
}

// --- Tariff serialisation for localStorage ----------------------------------

const TARIFF_PREFIX = 'tariff:';

export function tariffStorageKey(assetName) {
  return TARIFF_PREFIX + (assetName || 'default').replace(/[^a-zA-Z0-9._-]+/g, '_');
}

export function loadTariff(assetName) {
  try {
    const raw = localStorage.getItem(tariffStorageKey(assetName));
    if (!raw) return null;
    const parsed = JSON.parse(raw);
    return normalizeTariff(parsed);
  } catch (_) {
    return null;
  }
}

export function saveTariff(assetName, tariff) {
  try {
    localStorage.setItem(tariffStorageKey(assetName), JSON.stringify(tariff));
  } catch (_) {/* quota etc. */}
}

export function normalizeTariff(t) {
  return {
    currency: String(t?.currency ?? 'USD'),
    peakRate: Number(t?.peakRate ?? 0) || 0,
    offpeakRate: Number(t?.offpeakRate ?? 0) || 0,
    peakHours: Array.isArray(t?.peakHours)
      ? t.peakHours.map(([a, b]) => [Number(a) | 0, Number(b) | 0])
      : [],
  };
}

/**
 * Parse "09:00-21:00, 14:00-15:00" (HH:MM hour ranges) into peakHours.
 * Minutes are dropped for v0.4 (hour-of-day buckets).
 */
export function parsePeakHoursString(text) {
  if (!text || !text.trim()) return [];
  const out = [];
  for (const piece of text.split(',')) {
    const m = piece.trim().match(/^(\d{1,2})(?::\d{2})?\s*-\s*(\d{1,2})(?::\d{2})?$/);
    if (!m) continue;
    const a = parseInt(m[1], 10);
    const b = parseInt(m[2], 10);
    if (a >= 0 && a < 24 && b >= 0 && b < 24) out.push([a, b]);
  }
  return out;
}

export function peakHoursToString(peakHours) {
  return (peakHours ?? [])
    .map(([a, b]) => `${String(a).padStart(2, '0')}:00-${String(b).padStart(2, '0')}:00`)
    .join(', ');
}
